src/libcore/iter/range.rs

```diff
@@ -5,175 +5,408 @@ use crate::usize;
+    #[inline]
+    #[unstable(feature = "step_trait_ext", reason = "recently added", issue = "42168")]
+    fn successor(&self) -> Self {
+        self.forward(1).expect("overflow in `Step::successor`")
+    }
+
+    /// Returns the *successor* of `self`.
+    ///
+    /// If this would overflow the range of values supported by `Self`,
+    /// this method is defined to return the input value instead.
+    ///
+    /// # Invaraints
+    ///
+    /// For any `a` where `a.successor()` does not overflow:
+    ///
+    /// * `a.successor_saturating() == a.successor()`
+    ///
+    /// For any `a` where `a.successor()` does overflow:
+    ///
+    /// * `a.successor_saturating() == a`
+    #[inline]
+    #[unstable(feature = "step_trait_ext", reason = "recently added", issue = "42168")]
+    fn successor_saturating(&self) -> Self {
+        self.forward(1).unwrap_or_else(|| self.clone())
+    }
+
+    /// Returns the *successor* of `self`.
+    ///
+    /// # Safety
+    ///
+    /// It is undefined behavior if this operation exceeds the range of
+    /// values supported by `Self`. If you cannot guarantee that this
+    /// will not overflow, use `forward` or `successor` instead.
+    ///
+    /// For any `a`, if there exists `b` such that `b > a`,
+    /// it is safe to call `a.successor_unchecked()`.
+    #[inline]
+    #[unstable(feature = "unchecked_math", reason = "niche optimization path", issue = "none")]
+    unsafe fn successor_unchecked(&self) -> Self {
+        self.successor()
+    }
 
-    /// Adds one to this step, returning the result.
-    fn add_one(&self) -> Self;
+    /// Returns the value that would be obtained by taking the *predeccessor*
+    /// of `self` `count` times.
+    ///
+    /// Returns `None` if this would overflow the range of values supported by `Self`.
+    ///
+    /// # Invariants
+    ///
+    /// For any `a`, `n`, and `m` where `n + m` does not overflow:
+    ///
+    /// * `a.backward(n).and_then(|x| x.backward(m)) == a.backward (n + m)`
+    /// * `a.backward(n)` equals `Step::predecessor` applied to `a` `n` times
+    ///   * Corollary: `a.backward(0) == Some(a)`
+    /// * `a.backward(n).unwrap() <= a`
+    fn backward(&self, count: usize) -> Option<Self>;
 
-    /// Subtracts one to this step, returning the result.
-    fn sub_one(&self) -> Self;
+    /// Returns the *predecessor* of `self`.
+    ///
+    /// If this would underflow the range of values supported by `Self`,
+    /// this method is allowed to panic or wrap. Suggested behavior is
+    /// to panic when debug assertions are enabled, and wrap otherwise.
+    ///
+    /// # Invariants
+    ///
+    /// For any `a` where `a.predeccessor()` does not underflow:
+    ///
+    /// * `a == a.predecessor().successor()`
+    /// * `a.predecessor() == a.backward(1).unwrap()`
+    /// * `a.predecessor() <= a`
+    #[inline]
+    #[unstable(feature = "step_trait_ext", reason = "recently added", issue = "42168")]
+    fn predeccessor(&self) -> Self {
```

Comment:
*predecessor

here and at other places

Reply:
>.< I can't spell

fixed